backend/attachRendererFiber.js

```diff
@@ -0,0 +1,252 @@
+  // The naming is confusing.
+  // They deal with opaque nodes (fibers), not elements.
+  function getNativeFromReactElement(fiber) {
+    try {
+      return renderer.findHostInstanceByFiber(fiber);
+    } catch (err) {
+      // The fiber might have unmounted by now.
+      return null;
+    }
+  }
+  function getReactElementFromNative(node) {
+    const fiber = renderer.findFiberByHostInstance(node);
+    return fiber;
+  }
+
+  // This is a slightly annoying indirection.
+  // It is currently necessary because DevTools wants
+  // to use unique objects as keys for instances.
+  // However fibers have two versions.
+  // We use this set to remember first encountered fiber for
+  // each conceptual instance.
+  const opaqueNodes = new Set();
+  function getOpaqueNode(fiber) {
+    if (opaqueNodes.has(fiber)) {
+      return fiber;
+    }
+    const {alternate} = fiber;
+    if (alternate != null && opaqueNodes.has(alternate)) {
+      return alternate;
+    }
+    opaqueNodes.add(fiber);
+    return fiber;
+  }
+
+  function haveChildrenChanged(prevChildren, nextChildren) {
+    if (!Array.isArray(prevChildren) || !Array.isArray(nextChildren)) {
+      return prevChildren !== nextChildren;
+    }
+    if (prevChildren.length !== nextChildren.length) {
+      return true;
+    }
+    for (let i = 0; i < prevChildren.length; i++) {
+      if (prevChildren[i] !== nextChildren[i]) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  function hasDataChanged(prevData, nextData) {
+    return (
+      prevData.ref !== nextData.ref ||
+      prevData.source !== nextData.source ||
+      prevData.props !== nextData.props ||
+      prevData.state !== nextData.state ||
+      prevData.context !== nextData.context ||
+      prevData.text !== nextData.text ||
+      haveChildrenChanged(prevData.children, nextData.children)
+    );
+  }
+
+  let pendingEvents = [];
+
+  function flushPendingEvents() {
+    const events = pendingEvents;
+    pendingEvents = [];
+    for (let i = 0; i < events.length; i++) {
+      const event = events[i];
+      hook.emit(event._event, event);
+    }
+  }
+
+  function enqueueMount(fiber) {
+    pendingEvents.push({
+      // TODO: the naming is confusing. `element` is *not* a React element. It is an opaque ID.
+      element: getOpaqueNode(fiber),
+      data: getDataFiber(fiber, getOpaqueNode),
+      renderer: rid,
+      _event: 'mount',
+    });
+
+    const isRoot = fiber.tag === 3;
+    if (isRoot) {
+      pendingEvents.push({
+        element: getOpaqueNode(fiber),
+        renderer: rid,
+        _event: 'root',
+      });
+    }
+  }
+
+  function enqueueUpdateIfNecessary(fiber) {
+    const nextData = getDataFiber(fiber, getOpaqueNode);
+    const prevData = getDataFiber(fiber.alternate, getOpaqueNode);
+    // Avoid unnecessary updates since they are common
+    // when something changes deep in the tree due to setState.
+    if (!hasDataChanged(prevData, nextData)) {
```

Comment:
Would it be better to do this at the fiber level instead? Maybe it's more fragile to updates but faster since you don't have to create the other objects.